postgres/src/main/scala/zio/sql/postgresql/PostgresModule.scala

```diff
@@ -222,7 +158,92 @@ trait PostgresModule extends Jdbc { self =>
           buildExpr(on)
           val _ = builder.append(" ")
       }
-    buildReadString(read)
-    builder.toString()
+
+    def buildReadString[A <: SelectionSet[_]](read: self.Read[_]): Unit =
+      read match {
+        case read0 @ Read.Select(_, _, _, _, _, _, _, _) =>
+          object Dummy {
+            type F
+            type A
+            type B <: SelectionSet[A]
+          }
+          val read = read0.asInstanceOf[Read.Select[Dummy.F, Dummy.A, Dummy.B]]
+          import read._
+
+          builder.append("SELECT ")
+          buildSelection(selection.value)
+          builder.append(" FROM ")
+          buildTable(table)
+          whereExpr match {
+            case Expr.Literal(true) => ()
+            case _                  =>
+              builder.append(" WHERE ")
+              buildExpr(whereExpr)
+          }
+          groupBy match {
+            case _ :: _ =>
+              builder.append(" GROUP BY ")
+              buildExprList(groupBy)
+
+              havingExpr match {
+                case Expr.Literal(true) => ()
+                case _                  =>
+                  builder.append(" HAVING ")
+                  buildExpr(havingExpr)
+              }
+            case Nil    => ()
+          }
+          orderBy match {
+            case _ :: _ =>
+              builder.append(" ORDER BY ")
+              buildOrderingList(orderBy)
+            case Nil    => ()
+          }
+          limit match {
+            case Some(limit) =>
+              builder.append(" LIMIT ").append(limit)
+            case None        => ()
+          }
+          offset match {
+            case Some(offset) =>
+              val _ = builder.append(" OFFSET ").append(offset)
+            case None         => ()
+          }
+
+        case Read.Union(left, right, distinct) =>
+          buildReadString(left)
+          builder.append(" UNION ")
+          if (!distinct) builder.append("ALL ")
+          buildReadString(right)
+
+        case Read.Literal(values) =>
+          val _ = builder.append(" (").append(values.mkString(",")).append(") ") //todo fix needs escaping
+      }
+
+    def buildDeleteString(delete: self.Delete[_, _]): Unit = {
+      builder.append("DELETE FROM ")
+      buildTable(delete.table)
+      builder.append(" WHERE ")
+      buildExpr(delete.whereExpr)
+    }
+
+    def render(): String = builder.toString
+
+  }
+
+  object PostgresFunctionDef {
+    val Sind = FunctionDef[Double, Double](FunctionName("sind"))
+  }
+
+  override def renderDelete(delete: self.Delete[_, _]): String = {
+    val renderer = Renderer()
+    renderer.buildDeleteString(delete)
+    renderer.render()
```

Comment:
so if I did
```scala
val renderer = Renderer()
renderer.buildDeleteString(delete)
renderer.render()
renderer.buildReadString(read)
renderer.render()
```
I would get some invalid SQL in the end.
I think introducing this Renderer and two-step String rendering is overcomplicating things. Wouldn't top level `buildDeleteString` and `buildReadString` achieve essentially the same without the risk of invalid state? Or even implementing them directly in `renderDelete` and `renderRead`?
The common private functions can still be shared.

Reply:
Good point. Will fix it.